doc/function-composition-cheatsheet.md

```diff
@@ -0,0 +1,172 @@
+# Composition of Functions
+
+## Why Another Cheatsheet?
+
+FP is about composition of functions. And it is surprisingly hard to find a cheatsheet just for composition of functions in Haskell, so let's try to start with a simple one and iterate on it over time. PRs more than welcome :)
+
+## Change Logs:
+
+| Date | Contributors | Description |
+| ---  | ---          | ---
+|2019-09-24 | [Daniel Deng](https://github.com/sinogermany) | Initial version, Function as Functor, Applicative and Monad, Kleisli Arrow Composition
+
+## Special Thanks To:
+
+- [Shine Li](https://github.com/shineli1984) who helped me along the way while I worked on the initial version.
+
+## GHC Source Code
+
+<details>
+<summary>Function as Functor, Applicative and Monad</summary>
+
+```haskell
+-- | @since 2.01
+instance Functor ((->) r) where
+    fmap = (.)
+
+-- | @since 2.01
+instance Applicative ((->) a) where
+    pure = const
+    (<*>) f g x = f x (g x)
+    liftA2 q f g x = q (f x) (g x)
+
+-- | @since 2.01
+instance Monad ((->) r) where
+    f >>= k = \ r -> k (f r) r
+```
+
+</details>
+
+<details><summary>"Komposition" of Functions (Kleisli Arrows)</summary>
+
+```haskell
+-- | Left-to-right composition of Kleisli arrows.
+(>=>)       :: Monad m => (a -> m b) -> (b -> m c) -> (a -> m c)
+f >=> g     = \x -> f x >>= g
+
+-- | Right-to-left composition of Kleisli arrows. @('>=>')@, with the arguments
+-- flipped.
+--
+-- Note how this operator resembles function composition @('.')@:
+--
+-- > (.)   ::            (b ->   c) -> (a ->   b) -> a ->   c
+-- > (<=<) :: Monad m => (b -> m c) -> (a -> m b) -> a -> m c
+(<=<)       :: Monad m => (b -> m c) -> (a -> m b) -> (a -> m c)
+(<=<)       = flip (>=>)
+
+```
+
+</details>
+
+## Cheatsheet
+
+<details>
+<summary>Function as Functor (Function Composition)</summary>
+
+```haskell
+f x = f1 (f2 (f3 (fn x)))
+
+-- composition
+f = f1 . f2 . f3 . fn
+
+-- GHC.Base.Functor
+-- Usually we just use (.) though.
+f = f1 `fmap` f2 `fmap` f3 `fmap` fn
+f = f1 <$> f2 <$> f3 <$> fn
+
+-- Control.Arrow
+f = f1 <<< f2 <<< f3 <<< fn
+f = fn >>> f3 >>> f2 >>> f1
+
+-- Flow
+f = f1 <. f2 <. f3 <. fn
+f = fn .> f3 .> f2 .> f1
+```
+
+```haskell
+-- Functor Law
+f . id == f
+id . f == f
+```
+
+</details>
+
+<details><summary>Function as Applicative</summary>
+
+```haskell
+f x = f1 x (f2 x)
+f = f1 <*> f2
+```
+
+```haskell
+f x = g (f1 x) (f2 x) (f3 x) (fn x)
+
+-- Applicative Style
+f = g <$> f1 <*> f2 <*> f3 <*> fn
+f = g . f1 <*> f2 <*> f3 <*> fn
+--
+```
+
+```haskell
+f x = g (f1 x) (f2 x) (f3 x) (fn x) x
+
+-- Applicative Style
+f = g <$> f1 <*> f2 <*> f3 <*> fn <*> id
```

Comment:
@shineli1984 is it worth the effort of flipping for the purpose of removing the `id` here?

Reply:
Yes. A flipped g is usually what we expect when using <*>. Otherwise, we would use =<<

Reply:
Thanks @shineli1984 . 

Question though, if we have:
```haskell
f :: a -> b -> c -> d -> e
```

Then we would have:
```haskell
flip f :: b -> a -> c -> d -> e
```
So how do we solve this without using `id`?
```haskell
f x = g (f1 x) (f2 x) (f3 x) ... (fn x) x
```

Reply:
Sorry. I wasn’t clear. I meant we would expect g to have x as the first argument